client/shaders/nodes_shader/opengl_fragment.glsl

```diff
@@ -50,7 +62,42 @@ varying vec3 tsEyeVec;
 varying vec3 lightVec;
 varying vec3 tsLightVec;
 
+varying vec3 viewVec;
+
 #ifdef ENABLE_DYNAMIC_SHADOWS
+vec4 perm(vec4 x)
+{
+	return mod(((x * 34.0) + 1.0) * x, 289.0);
+}
+
+#if ((defined(MATERIAL_WAVING_LIQUID) && defined(ENABLE_WATER_REFLECTIONS)) || defined(ENABLE_BUMPMAPS))
+float snoise(vec3 p)
+{
+	vec3 a = floor(p);
+	vec3 d = p - a;
+	d = d * d * (3.0 - 2.0 * d);
+
+	vec4 b = a.xxyy + vec4(0.0, 1.0, 0.0, 1.0);
+	vec4 k1 = perm(b.xyxy);
+	vec4 k2 = perm(k1.xyxy + b.zzww);
+
+	vec4 c = k2 + a.zzzz;
+	vec4 k3 = perm(c);
+	vec4 k4 = perm(c + 1.0);
+
+	vec4 o1 = fract(k3 * (1.0 / 41.0));
+	vec4 o2 = fract(k4 * (1.0 / 41.0));
+
+	vec4 o3 = o2 * d.z + o1 * (1.0 - d.z);
+	vec2 o4 = o3.yw * d.x + o3.xz * (1.0 - d.x);
+
+	return o4.y * d.y + o4.x * (1.0 - d.y);
+}
+
+float wave_noise(vec3 p, float off) {
+	return snoise(p + vec3(0.0, 0.0, off)) * 0.4 + snoise(2.0 * p + vec3(0.0, off, off)) * 0.1 + snoise(3.0 * p + vec3(0.0, off, off)) * 0.075 + snoise(4.0 * p + vec3(-off, off, 0.0)) * 0.05;
```

Comment:
Is there a simpler way. We call snoise 4 times here, then later we call wave_noise 3 times.
In the vertex shader we're only calling snoise once to get the wave pattern.

Reply:
Good point, the 4 calls in wave_noise i know seem a bit much and it’s one more than would technically be needed for regular fractal noise - however, especially with these reflections, any artifacts in the noise show very clearly and these 4 just look significantly better. 
The fact wave_noise has to be called three times is because we need normal information in the fragment shader, which we don’t need in the vertex shader. I did implement it for the vertex shader in my „shader pack“ for 5.8.0 and that does almost definitely improve performance, but also shows very distinct artifacts.
With many functions you can get their gradient as an analytical function, but I‘m not sure how that would work with noise like this. I could look into using analytical functions like sine waves instead of noise, which would reduce the calls to effectively 2 calls to the „noise“ function, but it would introduce periodicity and I‘m not sure about the performance improvement.
If anyone else has ideas as to how to optimize this, let me know.